main/platform/hardened-js.md

```diff
@@ -0,0 +1,199 @@
+# Intro to Hardened JavaScript
+
+_Status: WIP_.
+
+## Preface: Fodder / Brainstorm
+
+ - structure from [E in a Walnut](http://www.skyhunter.com/marcs/ewalnut.html#SEC8)
+   - as adapted in [Practical Security: The Mafia game — Monte 0\.1 documentation](https://monte.readthedocs.io/en/latest/ordinary-programming.html)
+ - JSON / Justin / Jessie as in [Jessica](https://github.com/agoric-labs/jessica)
+ - Build slides with [Remark](https://remarkjs.com/#1)
+   - example: [kumc\-bmi/naaccr\-tumor\-data](https://github.com/kumc-bmi/naaccr-tumor-data)
+
+## Overview
+
+ - JavaScript is popular
+ - JavaScript has some messy parts
+ - Stick to Hardened JavaScript:
+   - Objects in Jessie
+   - Data in JSON
+   - Expressions in Justin
+   - Jessie details
+
+## Simple Objects
+
+Singleton, stateless:
+
+```js
+const origin = {
+    getX: () => 0,
+    getY: () => 0,
+};
+```
+
+```console
+> origin.getY()
+0
+```
+
+
+## Object makers
+
+```js
+const makeCounter = init => {
+  let value = init;
+  return {
+    increment: () => {
+      value += 1;
+      return value;
+    },
+    makeOffsetCounter: delta => makeCounter(value + delta),
+  };
+};
+```
+
+```console
+$ node
+Welcome to Node.js v14.16.0.
+Type ".help" for more information
+> const c1 = makeCounter(1);
+> c1.increment();
+2
+> const c2 = c1.makeOffsetCounter(10);
+> c1.increment();
+3
+> c2.increment();
+13
+> [c1.increment(), c2.increment()];
+[ 4, 14 ]
+```
+
+ - An object is a record of functions that close over shared state.
+ - A `makeCounter` function take the place of `class Counter`.
+ - `this` is not in Jessie.
+
+
+_TODO: `decrement()`, facets_
+
+## WARNING: Pervasive Mutability
+
+```console
+> c1.increment = () => { console.log('launch the missiles!'); }
+[Function (anonymous)]
+> c1.increment()
+launch the missiles!
+```
+
+
+## Defensive objects with `harden()`
+
+```js
+const makeCounter = init => {
+  let value = init;
+  return harden({
+    increment: () => {
+      value += 1;
+      return value;
+    },
+    makeOffsetCounter: delta => makeCounter(value + delta),
+  });
+};
+```
+
+```
+> const c3 = Object.freeze(c1.makeOffsetCounter(10));
+undefined
+> c3.increment = () => { console.log('launch the missiles!'); }
+TypeError: Cannot assign to read only property 'increment' of object '#<Object>'
+> c3.increment()
+15
+```
+
+ - _caveat_: exception is thrown in strict mode. REPL might not throw.
+ - regardless, the object defended itself
+ - Jessie pre-defines `harden` (_as does Agoric smart contract framework_)
+   - `npm install @agoric/harden` and `import harden from '@agoric/harden';` otherwise
```

Comment:
We really don't want people to use `@agoric/harden`, since ses is ready for this usage pattern.

```suggestion
   - see the `ses` [hardened Javascript package](https://github.com/endojs/endo/tree/master/packages/ses#harden) otherwise
```